Flaky TestCassandraTokenSplitManager.setUp

```
Error:  io.trino.plugin.cassandra.TestCassandraTokenSplitManager -- Time elapsed: 17.60 s <<< ERROR!
com.datastax.oss.driver.api.core.AllNodesFailedException: Could not reach any contact point, make sure you've provided valid addresses (showing first 1 nodes, use getAllErrors() for more): Node(endPoint=localhost/127.0.0.1:32770, hostId=null, hashCode=6ebf142c): [com.datastax.oss.driver.api.core.connection.ConnectionInitException: [s2|control|id: 0xd0529295, L:/127.0.0.1:58048 - R:localhost/127.0.0.1:32770] Protocol initialization request, step 1 (OPTIONS): unexpected failure (com.datastax.oss.driver.api.core.connection.ClosedConnectionException: Lost connection to remote peer)]
	at com.datastax.oss.driver.api.core.AllNodesFailedException.copy(AllNodesFailedException.java:141)
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures.getUninterruptibly(CompletableFutures.java:149)
	at com.datastax.oss.driver.api.core.session.SessionBuilder.build(SessionBuilder.java:835)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:181)
	at io.trino.plugin.cassandra.CassandraSession.close(CassandraSession.java:615)
	at io.trino.plugin.cassandra.CassandraServer.<init>(CassandraServer.java:117)
	at io.trino.plugin.cassandra.CassandraServer.<init>(CassandraServer.java:79)
	at io.trino.plugin.cassandra.CassandraServer.<init>(CassandraServer.java:73)
	at io.trino.plugin.cassandra.TestCassandraTokenSplitManager.setUp(TestCassandraTokenSplitManager.java:48)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:[184](https://github.com/trinodb/trino/actions/runs/7008402419/job/19067263055#step:5:185)3)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:[188](https://github.com/trinodb/trino/actions/runs/7008402419/job/19067263055#step:5:189))
	Suppressed: com.datastax.oss.driver.api.core.connection.ConnectionInitException: [s2|control|id: 0xd0529295, L:/127.0.0.1:58048 - R:localhost/127.0.0.1:32770] Protocol initialization request, step 1 (OPTIONS): unexpected failure (com.datastax.oss.driver.api.core.connection.ClosedConnectionException: Lost connection to remote peer)
		at com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler$InitRequest.fail(ProtocolInitHandler.java:356)
		at com.datastax.oss.driver.internal.core.channel.ChannelHandlerRequest.onFailure(ChannelHandlerRequest.java:104)
		at com.datastax.oss.driver.internal.core.channel.InFlightHandler.fail(InFlightHandler.java:381)
		at com.datastax.oss.driver.internal.core.channel.InFlightHandler.abortAllInFlight(InFlightHandler.java:371)
		at com.datastax.oss.driver.internal.core.channel.InFlightHandler.abortAllInFlight(InFlightHandler.java:353)
		at com.datastax.oss.driver.internal.core.channel.InFlightHandler.channelInactive(InFlightHandler.java:331)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:303)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)
		at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:411)
		at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:376)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:305)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:301)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)
		at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)
		at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:813)
		at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
		at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused by: com.datastax.oss.driver.api.core.connection.ClosedConnectionException: Lost connection to remote peer
	Suppressed: java.lang.NullPointerException: Cannot invoke "io.trino.plugin.cassandra.CassandraServer.close()" because "this.server" is null
		at io.trino.plugin.cassandra.TestCassandraTokenSplitManager.tearDown(TestCassandraTokenSplitManager.java:57)
		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
		at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1116)
		... 6 more
```

https://github.com/trinodb/trino/actions/runs/7009448726/job/19069391217
https://github.com/trinodb/trino/actions/runs/7008402419/job/19067263055

--- a/plugin/trino-cassandra/pom.xml
+++ b/plugin/trino-cassandra/pom.xml
@@ -242,6 +242,12 @@
 
         <dependency>
             <groupId>org.testcontainers</groupId>
+            <artifactId>cassandra</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.testcontainers</groupId>
             <artifactId>testcontainers</artifactId>
             <scope>test</scope>
         </dependency>

--- a/plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/CassandraServer.java
+++ b/plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/CassandraServer.java
@@ -18,20 +18,19 @@ import com.datastax.oss.driver.api.core.CqlSessionBuilder;
 import com.datastax.oss.driver.api.core.ProtocolVersion;
 import com.datastax.oss.driver.api.core.config.DriverConfigLoader;
 import com.datastax.oss.driver.api.core.config.ProgrammaticDriverConfigLoaderBuilder;
-import com.datastax.oss.driver.api.core.cql.ResultSet;
-import com.datastax.oss.driver.api.core.cql.Row;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.io.Resources;
 import io.airlift.json.JsonCodec;
 import io.airlift.log.Logger;
 import io.airlift.units.Duration;
 import io.trino.testing.ResourcePresence;
-import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.CassandraContainer;
+import org.testcontainers.containers.wait.CassandraQueryWaitStrategy;
+import org.testcontainers.utility.DockerImageName;
 
 import java.io.Closeable;
 import java.io.File;
 import java.io.IOException;
-import java.net.InetSocketAddress;
 import java.nio.file.Path;
 import java.util.List;
 import java.util.Map;
@@ -52,7 +51,7 @@ import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.MINUTES;
 import static java.util.concurrent.TimeUnit.NANOSECONDS;
 import static java.util.concurrent.TimeUnit.SECONDS;
-import static org.assertj.core.api.Assertions.assertThat;
+import static org.testcontainers.containers.CassandraContainer.CQL_PORT;
 import static org.testcontainers.utility.MountableFile.forHostPath;
 
 public class CassandraServer
@@ -60,11 +59,9 @@ public class CassandraServer
 {
     private static final Logger log = Logger.get(CassandraServer.class);
 
-    private static final int PORT = 9142;
-
     private static final Duration REFRESH_SIZE_ESTIMATES_TIMEOUT = new Duration(1, MINUTES);
 
-    private final GenericContainer<?> dockerContainer;
+    private final CassandraContainer<?> dockerContainer;
     private final CassandraSession session;
 
     public CassandraServer()
@@ -76,18 +73,18 @@ public class CassandraServer
     public CassandraServer(String imageName, String configFileName)
             throws Exception
     {
-        this(imageName, ImmutableMap.of(), "/etc/cassandra/cassandra.yaml", configFileName);
+        this(DockerImageName.parse(imageName), ImmutableMap.of(), "/etc/cassandra/cassandra.yaml", configFileName);
     }
 
-    public CassandraServer(String imageName, Map<String, String> environmentVariables, String configPath, String configFileName)
+    public CassandraServer(DockerImageName imageName, Map<String, String> environmentVariables, String configPath, String configFileName)
             throws Exception
     {
         log.debug("Starting cassandra...");
 
-        this.dockerContainer = new GenericContainer<>(imageName)
-                .withExposedPorts(PORT)
+        this.dockerContainer = new CassandraContainer<>(imageName)
                 .withCopyFileToContainer(forHostPath(prepareCassandraYaml(configFileName)), configPath)
                 .withEnv(environmentVariables)
+                .waitingFor(new CassandraQueryWaitStrategy())
                 .withStartupTimeout(java.time.Duration.ofMinutes(10));
         this.dockerContainer.start();
 
@@ -99,27 +96,15 @@ public class CassandraServer
         driverConfigLoaderBuilder.withStringList(METADATA_SCHEMA_REFRESHED_KEYSPACES, List.of());
 
         CqlSessionBuilder cqlSessionBuilder = CqlSession.builder()
-                .withApplicationName("TestCluster")
-                .addContactPoint(new InetSocketAddress(this.dockerContainer.getHost(), this.dockerContainer.getMappedPort(PORT)))
-                .withLocalDatacenter("datacenter1")
+                .addContactPoint(dockerContainer.getContactPoint())
+                .withLocalDatacenter(dockerContainer.getLocalDatacenter())
                 .withConfigLoader(driverConfigLoaderBuilder.build());
 
-        CassandraSession session = new CassandraSession(
+        this.session = new CassandraSession(
                 CASSANDRA_TYPE_MANAGER,
                 JsonCodec.listJsonCodec(ExtraColumnMetadata.class),
                 cqlSessionBuilder::build,
                 new Duration(1, MINUTES));
-
-        try {
-            checkConnectivity(session);
-        }
-        catch (RuntimeException e) {
-            session.close();
-            this.dockerContainer.stop();
-            throw e;
-        }
-
-        this.session = session;
     }
 
     private static String prepareCassandraYaml(String fileName)
@@ -152,16 +137,7 @@ public class CassandraServer
 
     public int getPort()
     {
-        return dockerContainer.getMappedPort(PORT);
-    }
-
-    private static void checkConnectivity(CassandraSession session)
-    {
-        ResultSet result = session.execute("SELECT release_version FROM system.local");
-        List<Row> rows = result.all();
-        assertThat(rows.size()).isEqualTo(1);
-        String version = rows.get(0).getString(0);
-        log.debug("Cassandra version: %s", version);
+        return dockerContainer.getMappedPort(CQL_PORT);
     }
 
     public void refreshSizeEstimates(String keyspace, String table)

--- a/plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestDatastaxConnectorSmokeTest.java
+++ b/plugin/trino-cassandra/src/test/java/io/trino/plugin/cassandra/TestDatastaxConnectorSmokeTest.java
@@ -15,6 +15,7 @@ package io.trino.plugin.cassandra;
 
 import com.google.common.collect.ImmutableMap;
 import io.trino.testing.QueryRunner;
+import org.testcontainers.utility.DockerImageName;
 
 import java.sql.Timestamp;
 import java.util.Map;
@@ -30,7 +31,7 @@ public class TestDatastaxConnectorSmokeTest
             throws Exception
     {
         CassandraServer server = closeAfterClass(new CassandraServer(
-                "datastax/dse-server:6.8.25",
+                DockerImageName.parse("datastax/dse-server:6.8.25").asCompatibleSubstituteFor("cassandra"),
                 Map.of(
                         "DS_LICENSE", "accept",
                         "DC", "datacenter1"),

--- a/plugin/trino-cassandra/src/test/resources/cassandra-dse.yaml
+++ b/plugin/trino-cassandra/src/test/resources/cassandra-dse.yaml
@@ -10,8 +10,6 @@ seed_provider:
       parameters:
           - seeds: "127.0.0.1"
 
-native_transport_port: 9142
-
 read_request_timeout_in_ms: 30000
 range_request_timeout_in_ms: 30000
 write_request_timeout_in_ms: 30000

--- a/plugin/trino-cassandra/src/test/resources/cu-cassandra-latest.yaml
+++ b/plugin/trino-cassandra/src/test/resources/cu-cassandra-latest.yaml
@@ -22,7 +22,6 @@ concurrent_counter_writes: 2
 concurrent_materialized_view_writes: 2
 
 listen_address: 127.0.0.1
-native_transport_port: 9142
 rpc_address: localhost
 broadcast_rpc_address: localhost
 

--- a/plugin/trino-cassandra/src/test/resources/cu-cassandra.yaml
+++ b/plugin/trino-cassandra/src/test/resources/cu-cassandra.yaml
@@ -260,8 +260,6 @@ ssl_storage_port: 7011
 listen_address: 127.0.0.1
 
 start_native_transport: true
-# port for the CQL native transport to listen for clients on
-native_transport_port: 9142
 
 # Whether to start the thrift rpc server.
 start_rpc: false

